cmd/kubeadm/app/cmd/phases/addons.go

```diff
@@ -17,186 +17,108 @@ limitations under the License.
+	Cfg() *kubeadmapi.InitConfiguration
+	Client() (clientset.Interface, error)
 }
 
-// EnsureAllAddons installs all addons to a Kubernetes cluster
-func EnsureAllAddons(cfg *kubeadmapi.InitConfiguration, client clientset.Interface) error {
-
-	addonActions := []func(cfg *kubeadmapi.InitConfiguration, client clientset.Interface) error{
-		dnsaddon.EnsureDNSAddon,
-		proxyaddon.EnsureProxyAddon,
+// NewAddonPhase returns the addon Cobra command
+func NewAddonPhase() workflow.Phase {
+	return workflow.Phase{
+		Name:     "addon",
+		Short:    "Installs required addons for passing Conformance tests",
+		CmdFlags: getAddonPhaseFlags("all"),
+		Phases: []workflow.Phase{
+			{
+				Name:     "coredns",
+				Short:    "Installs the CoreDNS addon to a Kubernetes cluster",
+				Long:     coreDNSAddonLongDesc,
+				CmdFlags: getAddonPhaseFlags(kubeadmconstants.CoreDNS),
+				Run:      runEnsureCoreDNSAddon,
+			},
+			{
+				Name:     "kube-proxy",
+				Short:    "Installs the kube-proxy addon to a Kubernetes cluster",
+				Long:     kubeProxyAddonLongDesc,
+				CmdFlags: getAddonPhaseFlags(kubeadmconstants.KubeProxy),
+				Run:      runEnsureKubeProxyAddon,
+			},
+		},
 	}
+}
 
-	glog.V(1).Infoln("[addon] installing all addons")
-	for _, action := range addonActions {
-		err := action(cfg, client)
-		if err != nil {
-			return err
-		}
+func getAddonData(c workflow.RunData) (*kubeadmapi.InitConfiguration, clientset.Interface, error) {
+	data, ok := c.(addonData)
+	if !ok {
+		return nil, nil, errors.New("addon phase invoked with an invalid data struct")
 	}
-
-	return nil
+	cfg := data.Cfg()
+	client, err := data.Client()
+	if err != nil {
+		return nil, nil, err
+	}
+	return cfg, client, err
 }
 
-// getAddonsSubCommands returns sub commands for addons phase
-func getAddonsSubCommands() []*cobra.Command {
-	cfg := &kubeadmapiv1beta1.InitConfiguration{}
-	// Default values for the cobra help text
-	kubeadmscheme.Scheme.Default(cfg)
-
-	var cfgPath, featureGatesString string
-	var subCmds []*cobra.Command
-	kubeConfigFile := kubeadmconstants.GetAdminKubeConfigPath()
-
-	subCmdProperties := []struct {
-		use      string
-		short    string
-		long     string
-		examples string
-		cmdFunc  func(cfg *kubeadmapi.InitConfiguration, client clientset.Interface) error
-	}{
-		{
-			use:      "all",
-			short:    "Installs all addons to a Kubernetes cluster",
-			long:     allAddonsLongDesc,
-			examples: allAddonsExample,
-			cmdFunc:  EnsureAllAddons,
-		},
-		{
-			use:     "coredns",
-			short:   "Installs the CoreDNS addon to a Kubernetes cluster",
-			long:    corednsAddonsLongDesc,
-			cmdFunc: dnsaddon.EnsureDNSAddon,
-		},
-		{
-			use:     "kube-proxy",
-			short:   "Installs the kube-proxy addon to a Kubernetes cluster",
-			long:    kubeproxyAddonsLongDesc,
-			cmdFunc: proxyaddon.EnsureProxyAddon,
-		},
+// runEnsureCoreDNSAddon installs CoreDNS addon to a Kubernetes cluster
+func runEnsureCoreDNSAddon(c workflow.RunData) error {
```

Comment:
not blocking.
I think that we can remove ensure (same for runEnsureKubeProxyAddon below)